src/tools/miri/tests/pass/uninhabited_not_read.rs

```diff
@@ -0,0 +1,24 @@
+//@ edition: 2021
+// In ed 2021 and below, we don't fallback `!` to `()`.
+// This would introduce a `! -> ()` coercion which would
+// be UB if we didn't disallow this explicitly.
```

Comment:
I don't understand the comment... does that mean we should test with two revisions, edition 2021 and 2024? Neither should be UB here, right?

Reply:
Not necessary to test in 2024 but I could add it just for completeness. 2024 should be equivalent to the explicitly ascribed `: !` case.

Reply:
So it seems that miri tests don't seem to allow editions and the fallback behavior doesn't *really* matter since I've got UI tests for that, and I'd rather not block this fix on tweaking miri testing infra lol.

Reply:
We usually set the edition via `@compile-flags: --edition <num>`, I think.

Reply:
Can you point me to a test that actually does that in the miri test suite? I didn't see any, and I got errors saying that the edition was being set twice when I tried that.

Reply:
Ah no, we set an edition for all tests by default. Yeah...
You can still add the fallback code as a test, right? Currently we use the 2021 edition so it does test the right thing. (And then we should have a Miri issue to track letting tests set the edition themselves.)

Reply:
OTOH [this](https://github.com/rust-lang/rust/blob/f7c938aaf0f497bb23e7338156435c3bf261e052/src/tools/miri/tests/ui.rs#L77) looks like ui_test understands editions so per-file and even per-revision edition should be possible. We are on a quite old version of ui_test though so maybe that's why it did not work?
Cc @oli-obk

Reply:
I cannot write the test that would force fallback to `()` since that's a compile error now. I added the test as-is from the issue, though.

Reply:
Anyways, I kinda don't want to tangle this up into this PR since I wrote sufficient UI tests to exercise this behavior, so perhaps file an issue to separate this? :D

Reply:
> I cannot write the test that would force fallback to `()` since that's a compile error now. I added the test as-is from the issue, though.

You added a test wanting edition 2021, and the edition Miri uses is 2021, so the test you originally intended should still be possible?

Reply:
Opened an issue: https://github.com/rust-lang/miri/issues/3860

Reply:
https://github.com/rust-lang/rust/pull/129392/files#diff-bad0427000dea00ae1fca50f869219d25310d8b2f4165c987e14041c4576e18aR102-R117

This is what I wanted to test.

Reply:
Nothing else needs to be done here unless you specifically want to see something exercised, otherwise I think we are just going in circles here lol

Reply:
FWIW once https://github.com/rust-lang/rust/pull/130467 lands, `//@edition` should work in Miri's ui tests, so if you want you can extend the test with revisions to run in both editions, or something like that.

Reply:
Cool, thanks